Event all the things!

calavera opened the pull request:
This is an attempt to make our events more rich and extensible.

In order to do that, we stop using `JSONMessage` as a struct to deliver events. That struct is an abomination for all kind of purposes, and the less we use it the better (imo). I created a specific purpose structure that includes new fields:
- Type: it's the type of resource the event is triggered for, `container`, `image`, `volume` and `network`.
- Action: it's the name of the event itself, `create`, `destroy`, `mount`, `connect` and so on.
- Actor: it's information about the resource, `ID`, `Name` and `Attributes`. Attributes is an open map to send extra metadata, like the labels for a container event.

**IMPORTANT**: This change is backwards compatible at the API level, but not at the CLI level.

The new event structure preserves old fields that we've been using until now for container and image events. People consuming the API won't see any difference in those events, but will start to see events without information in those fields, because they come from storage and network operations.

I've included a test to guarantee that we keep that backwards compatible information: https://github.com/docker/docker/compare/master...calavera:event_types?expand=1#diff-1e68c88f5c9fa605c0f84199cd04feb0R38

The CLI output has been completely modified and it's standard for all types of events, the syntax is as follow:

```
timestamp actor_type action actor_id (<optional>actor_attributes)
```

The attributes are listed in KEY=VALUE format to be consistent with other places where we show this kind of values, like labels.
## Some examples:
### Networking and container events

Network events that operate on containers (connect and disconnect) include the container id as part of the attributes:

```
2015-12-23T21:38:24.705709133Z network create 8b111217944ba0ba844a65b13efcd57dc494932ee2527577758f939315ba2c5b (name=test-event-network-local, type=bridge)
2015-12-23T21:38:25.075955890Z container create b4be644031a3d90b400f88ab3d4bdf4dc23adb250e696b6328b85441abe2c54e (name=test-network-container, image=busybox)
2015-12-23T21:38:25.119625123Z network connect 8b111217944ba0ba844a65b13efcd57dc494932ee2527577758f939315ba2c5b (name=test-event-network-local, container=b4be644031a3d90b400f88ab3d4bdf4dc23adb250e696b6328b85441abe2c54e, type=bridge)
2015-12-23T21:38:25.121051956Z container start b4be644031a3d90b400f88ab3d4bdf4dc23adb250e696b6328b85441abe2c54e (name=test-network-container, image=busybox)
```
### Storage and container events:

Storage events that operate on containers include the container id as part of the attributes:

```
2015-12-23T21:05:28.136212689Z volume create test-event-volume-local (driver=local)
2015-12-23T21:05:28.383462717Z volume mount test-event-volume-local (read/write=true, container=562fe10671e9273da25eed36cdce26159085ac7ee6707105fd534866340a5025, destination=/foo, driver=local, propagation=rprivate)
2015-12-23T21:05:28.650314265Z volume unmount test-event-volume-local (container=562fe10671e9273da25eed36cdce26159085ac7ee6707105fd534866340a5025, driver=local)
2015-12-23T21:05:28.716218405Z volume destroy test-event-volume-local (driver=local)
```
## Future improvements out of the scope here
1. Add a format flag with a go template.
## TODO
- [x] Modify the events documentation.

/cc @coolljt0725, @dnephin

- Fixes #18164
- Fixes #13845
- Fixes #18453 
- Fixes #18711
- Fixes #18674
- Fixes #18673
- Fixes #18672

coolljt0725 committed:
Add tests for docker events -f container.

Signed-off-by: Lei Jitang <leijitang@huawei.com>
(cherry picked from commit 531ecf59f5bd92b12d4548617ca7bf179c8179a3)

calavera committed:
Add event types.

- Stop serializing JSONMessage in favor of events.Message.
- Keep backwards compatibility with JSONMessage for container events.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Add volume events.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Add network events.

Signed-off-by: David Calavera <david.calavera@gmail.com>

thaJeztah commented:
Nice! Looking at the command line output, should we put the event earlier on? the type of event is kinda hurried now.  E.g.

```
2015-12-23T21:38:24.705709133Z network create 8b111217944ba0ba844a65b13efcd57dc494932ee2527577758f939315ba2c5b (test-event-network-local) (type: bridge)
2015-12-23T21:38:25.075955890Z container create b4be644031a3d90b400f88ab3d4bdf4dc23adb250e696b6328b85441abe2c54e (test-network-container) (image: busybox)
2015-12-23T21:38:25.119625123Z network connect 8b111217944ba0ba844a65b13efcd57dc494932ee2527577758f939315ba2c5b (test-event-network-local) (container: b4be644031a3d90b400f88ab3d4bdf4dc23adb250e696b6328b85441abe2c54e, type: bridge)
2015-12-23T21:38:25.121051956Z container start b4be644031a3d90b400f88ab3d4bdf4dc23adb250e696b6328b85441abe2c54e (test-network-container)  (image: busybox)
```

Or 

```
2015-12-23T21:38:24.705709133Z create network 8b111217944ba0ba844a65b13efcd57dc494932ee2527577758f939315ba2c5b (test-event-network-local) (type: bridge)
2015-12-23T21:38:25.075955890Z create container  b4be644031a3d90b400f88ab3d4bdf4dc23adb250e696b6328b85441abe2c54e (test-network-container) (image: busybox)
2015-12-23T21:38:25.119625123Z connect network 8b111217944ba0ba844a65b13efcd57dc494932ee2527577758f939315ba2c5b (test-event-network-local) (container: b4be644031a3d90b400f88ab3d4bdf4dc23adb250e696b6328b85441abe2c54e, type: bridge)
2015-12-23T21:38:25.121051956Z start container b4be644031a3d90b400f88ab3d4bdf4dc23adb250e696b6328b85441abe2c54e (test-network-container)  (image: busybox)
```

We could even consider a `--format` flag in the future

calavera commented:
Obligatory regular expression meme (do not look at `events_utils.go`):

![image](https://cloud.githubusercontent.com/assets/1050/11986057/307292e6-a980-11e5-80df-496255f180b8.png)

@thaJeztah I'm okay with moving the action at the beginning, let's wait to hear from other people to see if we agree in a format and I only need to change it once :smile_cat:

thaJeztah commented:
@calavera haha, absolutely, just my first impression looking at it. great work otherwise, I like it!

thaJeztah commented:
Oh, and
/cc @TomasTomecek @samsabed

dnephin commented on daemon/daemon.go:
```diff
@@ -1042,7 +1033,12 @@ func (daemon *Daemon) TagImage(newTag reference.Named, imageName string) error {
+		ID:   imageID.String(),
+		Name: newTag.String(),
+	}
+	daemon.EventsService.Log("tag", eventtypes.ImageEventType, actor)
```
I believe all the `eventtypes.ImageEventType` events should include the image labels in `Actor.Attributes`. Previously this was being injected by `daemon.GetLabels()`

calavera replied:
fixed, thanks!

dnephin replied:
I think they're all done except for this one place?

calavera replied:
:pensive: fixing

calavera replied:
actually there are a couple more places in pull/push :goberserk:

dnephin commented on daemon/daemon.go:
```diff
@@ -549,20 +549,11 @@ func (daemon *Daemon) GetByName(name string) (*container.Container, error) {
-			filter.Add("container", c.ID)
-		}
-	}
-	return events.NewFilter(filter, daemon.GetLabels)
```
I think this method `getEventFilter()` and `daemon.GetLabels()` can both be delete now.

calavera replied:
done! :boom:

dnephin commented:
Nice! I think this is awesome. We'll be able to make use of the new event data in the soon-be-released `docker-compose events` command.

calavera commented:
@thaJeztah one thing that I don't like about your suggested syntax is that it becomes harder to parse, because both name and attributes are optional at the end and are between parenthesis. But I'm open to suggestions.

calavera commented:
> But I'm open to suggestions.

maybe we should add the name to the attributes. That would leave us only with one option field, which I think it's better:

```
timestamp type action id (<optional>attributes)
```

Example:
```
2015-12-23T21:38:25.075955890Z container create b4be644031a3d90b400f88ab3d4bdf4dc23adb250e696b6328b85441abe2c54e (name: test-network-container, image: busybox)
```

TomasTomecek commented:
This is really impressive. Very nice work! A couple questions.

Will there be a precise documentation of what are the possible values (states) to expect? Regarding the new changes to layering, image IDs 'n stuff, would it make sense to also include digest? How about untagging an image, will I get the tag name? And last thing: would be super cool to unite events status and `inspect` status (`inspect` status contains text starting with capital letter and human-readable time which is a nightmare to parse).

thaJeztah commented:
@calavera oh, I like the suggestion of adding name to attributes. In addition, we can still add `--format` in a future PR. My primary concern was that the "type" of event got a bit lost in the other information, putting it more to the start of the line makes that easier to read (visually).

calavera commented:
> Regarding the new changes to layering, image IDs 'n stuff, would it make sense to also include digest? How about untagging an image, will I get the tag name?

@TomasTomecek yes, with the new attributes, we can include more relevant information about the events, like that one. This is not a one time only change, we can improve the information we add to each event in the future.

> And last thing: would be super cool to unite events status and inspect status (inspect status contains text starting with capital letter and human-readable time which is a nightmare to parse).

I'm not completely sure that's related to this change. I think we should discuss it further, but I understand your issues.

calavera committed:
Add filter by event type and documentation.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera commented:
@thaJeztah I've changed the log format to put the name in the attributes.

I've also added a new feature to be able to filter by event type with `--filter type=eventType`, like `--filter type=container` or `--filter type=network`.

Can I have some :+1: / :-1: to all this before moving it to code review?

@TomasTomecek documentation is coming.

vdemeester commented:
Really nice ! :+1: :wink:

thaJeztah commented:
@calavera thanks!

- moving name to attributes: lgtm
- filter by "type": lgtm

One think I'm still thinking about is if "type" is the proper name for this. My first impression would be that "type" would be "start", "stop", "build". Not sure what we would use for "container" / "image" (source? origin? class?)

Perhaps someone else has suggestions here for a better naming.

vdemeester commented:
@thaJeztah hum, I thought `type` was good because it follows the `type` of `inspect` though :stuck_out_tongue:.

calavera commented:
> My first impression would be that "type" would be "start", "stop", "build". Not sure what we would use for "container" / "image" (source? origin? class?)

That could be an interpretation, I'm going to give you another one :smile: 

The information a event holds does not describe the event itself, it describes the change of behaviors an actor(something inside docker) is exposed to. `type` does NOT mean `event type`, although I've made the mistake of naming several variables `eventType` myself, and I'll properly correct it soon. `type` means `actor type`, it's the resource docker performs an action with. A question that this interpretation raises is if the type should then be inside the `Actor` structure, that's sounds reasonable, and I'm okay either way.

Docker performs actions over the resources that it manages, therefore, what you consider types are labeled as `action`. Those actions an logged in the event log, but again, the events do NOT describe themselves, they describe the actions that docker performs.

thaJeztah commented:
> That could be an interpretation, I'm going to give you another one
> Docker performs actions over the resources that it manages, therefore, what you consider types are labeled as `action`

Guess that makes sense as well. Alright, let's stick with your implementation, "+1" on design for me.

aboch commented on daemon/container_operations_unix.go:
```diff
@@ -718,6 +722,11 @@ func (daemon *Daemon) DisconnectFromNetwork(container *container.Container, n li
+	daemon.LogNetworkEventWithAttributes(n, "disconnect", attributes)
 	return nil
 }
 
```
In this file can you please also take care of
`func (daemon *Daemon) releaseNetwork(container *Container)`

Given it is for a forced sandbox removal, I think we should generate the disconnect event from each connected network with something like:
```
+	attributes := map[string]string{
+		"container": container.ID,
+	}
+
 	for n := range networks {
+		if nw, err := daemon.FindNetwork(n); err == nil {
+                     daemon.LogNetworkEventWithAttributes(n, "disconnect", attributes)
+		}
 		networks[n] = &network.EndpointSettings{}
 	}
```

calavera replied:
sounds good, thanks.

calavera replied:
@aboch I think it's very weird to raise a disconnection event here when the container has not even connected to the network yet. As far as I understand this code, this is only a precaution to cleanup the sanbox before the container connects to a network, isn't it?

aboch replied:
@calavera Actually the `releaseNetwork()` is called by the monitor process every time the container stops.
And also on container start failure, but only if the network connect was successful.

In either case the logging would happen only for the networks the container was actually connected to.

calavera replied:
yeah, the fact that the caller is in `daemon/start.go` always confuses me. Thank you.

icecrime commented:
```
FAIL: docker_cli_events_unix_test.go:259: DockerSuite.TestEventsFilterVolumeAndNetworkType

docker_cli_events_unix_test.go:269:
    c.Assert(events, checker.HasLen, 2)
... obtained []string = []string{"2015-12-30T05:38:48.136240021Z network create a8dd34c70483e89271d405328194956131dbf92b109c291cbc42945d9bff8bf1 (name=test-event-network-local, type=bridge)", "2015-12-30T05:38:48.326981430Z network destroy a8dd34c70483e89271d405328194956131dbf92b109c291cbc42945d9bff8bf1 (name=test-event-network-local, type=bridge)", "2015-12-30T05:38:49.134810430Z network create 940a7b517e2d70a719ac6cc1c9954bade533e54d80487bcbcaaaf2ed9f41b0fd (type=bridge, name=test-event-network-type)", "2015-12-30T05:38:49.163590109Z volume create test-event-volume-type (driver=local)"}
... n int = 2
```

thaJeztah commented:
Looking at the discussion above, I think we're ready to move this to code review (and fix the failing tests during that)

duglin commented:
Just now getting to look at this, sorry for the delay. 
SGTM
As for the order, I prefer what is proposed - resource type being acted upon first, then the action. I think starting with something more abstract and getting more specific as we move to the right in the list is the easiest to deal with.  Of course, then actor_id should come before the action but the problem with that is its then a little harder to parse/grep when you just want to find "network create" it would need to be "network .* create " - not a huge issue but the human readability is better this way.

jfrazelle commented:
LGTM

prologic commented:
Will this break tools that rely on the current structure events from the Docker Events API such as [autodock](https://github.com/prologic/autodock) which uses [docker-py](https://pypi.python.org/pypi/docker-py)?

calavera commented:
@prologic I don't think there is a solid "yes" or "no" answer to that, it depends on how your project uses the events api. It might probably need some adjustments. Check what I wrote in the description about backwards compatibility:

![image](https://cloud.githubusercontent.com/assets/1050/12076685/85454fce-b16a-11e5-834c-397a8a0108a4.png)

Let me know if you have further questions about it.

prologic commented:
@calavera Thanks for the clarification! I'll keep an eye out for this one as autodock is a consumer of the Docker Events API. I've already had to adjust for new events in Docker 1.9 as well as some internal structural changes in the data from inspecting container configurations. Thanks again!

jfrazelle commented:
LGTM

LK4D4 commented:
LGTM

moxiegirl commented:
Was it likely customers see any visible changes in output --- for example if they were crunching output from the API --- do we need to mention something as being deprecated?

Otherwise, LGTM.

calavera commented:
good point @moxiegirl. The API still works with old fields, but I agree that's worth noticing in the deprecation document. I added it already.

thaJeztah commented:
docs LGTM

@moxiegirl the API is backwards compatible, but CLI output changed, I think we should mention that in the release notes

moxiegirl commented:
@thaJeztah  If by release notes you mean the changelog. Then we are good. :-)

thaJeztah commented:
potatoes potatos, yup that's what I meant, haha

calavera commented:
![](http://gph.is/1b9eRBa)

TomasTomecek commented:
@calavera very nice! can't wait to play with this

thaJeztah commented:
@TomasTomecek for playing; https://master.dockerproject.org is your friend :-) And please do! We can never have enough testers :)
